### ISSUE TITLE
Ship design can crash when right-clicking slots while dragging parts

### ISSUE THREAD

**Issue by rriegs** [freeorion/freeorion#1639]
### Environment

* **FreeOrion Version**: FreeOrionCH v0.4.7+ [build 2017-07-19.79efb88] CMake
* **Operating System**: Ubuntu 16.04.2 LTS (Linux 4.4.0-83 x86_64)
* **Graphic card used**: Intel Corporation 3rd Gen Core processor Graphics Controller (rev 09)
* **Fetched as**
  * Compiled from source

### Description

In the Design screen, left-click-dragging parts from the Ship Parts window over ship slots but then right-clicking (while still dragging) results in strange behavior: the part is placed, but its listing in the Ship Parts window vanishes and can no longer be selected.  Closing and reopening the Design screen restores the Ship Parts window's contents.

Further, left-click-dragging parts from *other ship slots* over ship slots and then right-clicking sometimes results in a segfault.  Steps to reproduce are as follows:

1. Place a ship part as normal.
2. Left-click-drag the placed ship part over another ship slot (empty or otherwise).
3. Right-click while still dragging.

The above procedure sometimes does not result in a crash, especially if ship parts are initially placed via right-click-while-dragging instead of as normal (as per Step 1 above).  Regardless, even if a crash does not occur, subsequent normal modifications to ship parts (left-click drag-and-dropping only) can later result in a crash.

I hypothesize that the ship part is somehow being placed in two ship slots at the same time, and that modifications that would normally delete the part from one slot then result in a dead pointer in the other slot, causing a crash.  Nothing useful seems to end up in the logs.

### Expected Result

At minimum, right-clicking while left-click-dragging a ship part should have no effect.

*Feature*: On the other hand, I think it would be neat if it placed a copy of the part while *still* dragging the part, thereby allowing you to place many copies of the same part in adjacent slots by left-click-dragging a part (from the Ship Parts window *or* from another slot) and then right-clicking multiple slots in sequence.

**Comment by dbenage-cx** [comment]
For the second portion, could only replicate exact steps without gdb. [abrt stacktrace](https://github.com/freeorion/freeorion/files/1164106/designwnd-part-drag-rclick.stacktrace.txt)

With gdb, added a step of double clicking the new part location.
Thus:
* Double click medium basic hull
* Double click standard plate armor
* Left button drag the armor part from top external slot to bottom external slot
* Still holding left button, right click
* Release left button
* Double left click bottom part slot

[backtrace](https://github.com/freeorion/freeorion/files/1164065/designwnd-part-drag-rclick.bt.txt)

**Comment by Vezzra** [comment]
Confirmed on OSX as reported in OP. Have been able to reproduce a crash with the `lldb` debugger attached. Crash occurred when dragging a part over a slot, before trying to drop it. Backtrace:
```
* thread #1: tid = 0x1008a, 0x0011d619 FreeOrion`SlotControl::DragDropEnter(GG::Pt const&, std::__1::map<GG::Wnd const*, bool, std::__1::less<GG::Wnd const*>, std::__1::allocator<std::__1::pair<GG::Wnd const* const, bool> > >&, GG::Flags<GG::ModKey>) + 109, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x31)
  * frame #0: 0x0011d619 FreeOrion`SlotControl::DragDropEnter(GG::Pt const&, std::__1::map<GG::Wnd const*, bool, std::__1::less<GG::Wnd const*>, std::__1::allocator<std::__1::pair<GG::Wnd const* const, bool> > >&, GG::Flags<GG::ModKey>) + 109
    frame #1: 0x005d8e49 FreeOrion`GG::Wnd::HandleEvent(GG::WndEvent const&) + 1307
    frame #2: 0x00579984 FreeOrion`GG::GUI::CheckedGetWindowUnder(GG::Pt const&, GG::Flags<GG::ModKey>) + 1518
    frame #3: 0x0057cb59 FreeOrion`GG::GUIImpl::HandleMouseMove(GG::Flags<GG::ModKey>, GG::Pt const&, GG::Pt const&, int) + 51
    frame #4: 0x0057e724 FreeOrion`GG::GUI::HandleGGEvent(GG::GUI::EventType, GG::Key, unsigned int, GG::Flags<GG::ModKey>, GG::Pt const&, GG::Pt const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const*) + 480
    frame #5: 0x005bf8dd FreeOrion`GG::SDLGUI::HandleSystemEvents() + 671
    frame #6: 0x002022b6 FreeOrion`HumanClientApp::HandleSystemEvents() + 46
    frame #7: 0x00203aaa FreeOrion`non-virtual thunk to HumanClientApp::HandleSystemEvents() + 22
    frame #8: 0x005570e6 FreeOrion`GG::ModalEventPump::operator()() + 52
    frame #9: 0x005c0080 FreeOrion`GG::SDLGUI::Run() + 46
    frame #10: 0x000b5d0d FreeOrion`mainSetupAndRun() + 591
    frame #11: 0x0035f9f6 FreeOrion`-[GUIController applicationDidFinishLaunching:] + 20
    frame #12: 0x9ac8d0f3 Foundation`__57-[NSNotificationCenter addObserver:selector:name:object:]_block_invoke + 50
    frame #13: 0x93eefc34 CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
    frame #14: 0x93dcf901 CoreFoundation`_CFXNotificationPost + 3713
    frame #15: 0x9ac6e2c4 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 92
    frame #16: 0x988c2868 AppKit`-[NSApplication _postDidFinishNotification] + 365
    frame #17: 0x988c2558 AppKit`-[NSApplication _sendFinishLaunchingNotification] + 239
    frame #18: 0x988bf05d AppKit`-[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 827
    frame #19: 0x988be92a AppKit`-[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 277
    frame #20: 0x9b0d5119 libobjc.A.dylib`-[NSObject performSelector:withObject:withObject:] + 84
    frame #21: 0x9aca4e57 Foundation`__76-[NSAppleEventManager setEventHandler:andSelector:forEventClass:andEventID:]_block_invoke + 118
    frame #22: 0x9aca499f Foundation`-[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 451
    frame #23: 0x9aca4794 Foundation`_NSAppleEventManagerGenericHandler + 211
    frame #24: 0x9bbea94e AE`aeDispatchAppleEvent(AEDesc const*, AEDesc*, unsigned long, unsigned char*) + 583
    frame #25: 0x9bbb5467 AE`dispatchEventAndSendReply(AEDesc const*, AEDesc*) + 44
    frame #26: 0x9bbb535a AE`aeProcessAppleEvent + 313
    frame #27: 0x94bdf1c1 HIToolbox`AEProcessAppleEvent + 55
    frame #28: 0x988b82ee AppKit`_DPSNextEvent + 2707
    frame #29: 0x988b71e5 AppKit`-[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 350
    frame #30: 0x988abb9c AppKit`-[NSApplication run] + 907
    frame #31: 0x98820fa0 AppKit`NSApplicationMain + 2082
    frame #32: 0x000b9681 FreeOrion`main + 719
    frame #33: 0x9420b6d9 libdyld.dylib`start + 1
```

**Comment by LGM-Doyle** [comment]
Thanks @rriegs for the perfect bug report, clear, precise and detailed.  I was able to replicate the problem(s).
Thanks @dbenage-cx and @Vezzra for the triage with backtraces.  

The crashes are caused by double deletes or dereferencing of the deleted/nullptr for the `PartControl` widget.  They are good examples of the type of error prevented by [PR #1632 converting GG to use smart pointers](https://github.com/freeorion/freeorion/pull/1632), which does not experience the crashes.

The underlying problem is that the second right mouse click starts a second drag drop process which deletes the `PartControl`. Subsequently, it is possible to derefernce the deleted pointer, or delete it a second time, either of which causes a crash.  I've produces a patch which fixes the problem.  It is based on #1632 and will have to be re-done if #1632 is not merged.  

I like the idea of right clicking doing a repeated drop.  I added a feature request #1640 to that effect.
